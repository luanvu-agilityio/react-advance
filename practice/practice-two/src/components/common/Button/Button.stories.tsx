import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import Button from './Button'

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: `

## CSS Variables Used

The button uses these variables from theme.css:

| Variable | Description |
|----------|-------------|
| \`--green-color-default\` | Primary green color for solid buttons |
| \`--green-shade-1\` | Darker green for hover states |
| \`--green-shade-2\` | Used for outlines and focus states |
| \`--black-color-default\` | Default text color |
| \`--black-shade-3\` | Medium gray color |
| \`--black-shade-5\` | Light gray background |
| \`--white-color\` | White text color |
| \`--font-family-primary\` | Main font family |
| \`--font-weight-bold\` | Font weight for button text |
`,
      },
    },
  },
  argTypes: {
    variant: {
      description: 'Visual style variant of the button',
      control: {
        type: 'select',
        options: ['solid', 'soft', 'outline', 'ghost'],
      },
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: 'solid' },
      },
    },
    size: {
      description: 'Size of the button',
      control: {
        type: 'select',
        options: ['1', '2', '3'],
      },
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: '2' },
      },
    },

    onClick: { action: 'clicked' },
    children: {
      description: 'The content of the button',
      control: 'text',
    },
  },
}

export default meta
type Story = StoryObj<typeof Button>

/**
 * Default button with standard settings
 */
export const Default: Story = {
  args: {
    children: 'Default Button',
    onClick: action('button clicked'),
  },
}

/**
 * Solid variant has a filled background with white text.
 *
 * CSS from theme.css:
 * ```css
 * .rt-variant-solid {
 *   background-color: var(--green-color-default);
 *   border: 2px solid var(--black-shade-1);
 *   color: var(--white-color);
 *   padding: 6px 12px;
 *   min-height: 36px;
 * }
 * ```
 */
export const Solid: Story = {
  args: {
    variant: 'solid',
    children: 'Solid Button',
    onClick: action('solid button clicked'),
  },
}

/**
 * Soft variant has no background or border, only text styling.
 *
 * CSS from theme.css:
 * ```css
 * .rt-variant-soft {
 *   color: var(--black-color-default);
 *   border: none;
 * }
 * ```
 */
export const Soft: Story = {
  args: {
    variant: 'soft',
    children: 'Soft Button',
    onClick: action('soft button clicked'),
  },
}

/**
 * Outline variant has a transparent background with a border.
 *
 * CSS from theme.css:
 * ```css
 * .rt-variant-outline {
 *   background-color: transparent;
 *   color: var(--black-color-default);
 *   border: 2px solid var(--green-shade-2);
 *   padding: 12px 16px;
 * }
 * ```
 */
export const Outline: Story = {
  args: {
    variant: 'outline',
    children: 'Outline Button',
    onClick: action('outline button clicked'),
  },
}

/**
 * Ghost variant has a light background that gets darker on hover.
 *
 * CSS from theme.css:
 * ```css
 * .rt-variant-ghost {
 *   background-color: var(--black-shade-5);
 *   color: var(--black-color-default);
 *   border: none;
 *   padding: 12px 16px;
 * }
 * ```
 */
export const Ghost: Story = {
  args: {
    variant: 'ghost',
    children: 'Ghost Button',
    onClick: action('ghost button clicked'),
  },
}

/**
 * Small buttons use size "1" with minimal padding.
 *
 * CSS from theme.css:
 * ```css
 * .rt-r-size-1 {
 *   padding: 6px 12px;
 * }
 * ```
 */
export const Small: Story = {
  args: {
    size: '1',
    children: 'Small Button',
    onClick: action('small button clicked'),
  },
}

/**
 * Medium buttons use size "2" with standard padding.
 *
 * CSS from theme.css:
 * ```css
 * .rt-r-size-2 {
 *   padding: 12px 16px;
 * }
 * ```
 */
export const Medium: Story = {
  args: {
    size: '2',
    children: 'Medium Button',
    onClick: action('medium button clicked'),
  },
}

/**
 * Large buttons use size "3" with extra padding.
 *
 * CSS from theme.css:
 * ```css
 * .rt-r-size-3 {
 *   padding: 12px 24px;
 * }
 * ```
 */
export const Large: Story = {
  args: {
    size: '3',
    children: 'Large Button',
    onClick: action('large button clicked'),
  },
}

/**
 * Disabled buttons have reduced opacity and cannot be clicked.
 */
export const Disabled: Story = {
  args: {
    disabled: true,
    children: 'Disabled Button',
    onClick: action('this should not be called'),
  },
}

/**
 * Buttons can contain icons alongside text.
 */
export const WithIcon: Story = {
  args: {
    children: (
      <>
        <span style={{ marginRight: '8px' }}>üîç</span>
        Search
      </>
    ),
    onClick: action('icon button clicked'),
  },
}

/**
 * A group showing all button variants for comparison.
 */
export const VariantComparison: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '16px', flexWrap: 'wrap' }}>
      <Button variant="solid" onClick={() => console.log('Button clicked')}>
        Solid
      </Button>
      <Button variant="soft" onClick={() => console.log('Button clicked')}>
        Soft
      </Button>
      <Button variant="outline" onClick={() => console.log('Button clicked')}>
        Outline
      </Button>
      <Button variant="ghost" onClick={() => console.log('Button clicked')}>
        Ghost
      </Button>
    </div>
  ),
}

/**
 * A group showing all button sizes for comparison.
 */
export const SizeComparison: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '16px', alignItems: 'center' }}>
      <Button size="1" onClick={() => console.log('Button clicked')}>
        Small
      </Button>
      <Button size="2" onClick={() => console.log('Button clicked')}>
        Medium
      </Button>
      <Button size="3" onClick={() => console.log('Button clicked')}>
        Large
      </Button>
    </div>
  ),
}
