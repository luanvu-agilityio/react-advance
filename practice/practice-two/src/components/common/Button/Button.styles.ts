import type { ButtonProps } from './Button'
import styled, { css } from 'styled-components'

export const BaseButton = styled.button<ButtonProps>`
  font-family: var(--font-family-primary);
  font-weight: var(--font-weight-bold);
  font-size: 15px;
  border-radius: 12px;
  cursor: pointer;
  transition: all 0.2s ease;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  outline: none;

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  /* Variant styles */
  ${(props) =>
    props.variant === 'solid' &&
    css`
      background-color: var(--green-color-default);
      border: 2px solid var(--green-shade-1);
      color: var(--white-color);

      &:hover:not(:disabled) {
        background-color: var(--green-shade-1);
      }

      &:focus {
        box-shadow: 0 0 0 2px var(--green-shade-2);
      }
    `}

  ${(props) =>
    props.variant === 'soft' &&
    css`
      color: var(--black-color-default);
      border: none;

      &:hover:not(:disabled) {
        color: var(--green-color-default);
      }

      &:focus {
        box-shadow: 0 0 0 2px var(--black-shade-3);
      }
    `}

  ${(props) =>
    props.variant === 'outline' &&
    css`
      background-color: transparent;
      color: var(--black-color-default);
      border: 2px solid var(--green-shade-2);

      &:hover:not(:disabled) {
        background-color: var(--green-shade-2);
        color: var(--white-color);
      }

      &:focus {
        box-shadow: 0 0 0 2px var(--green-shade-1);
      }
    `}

  ${(props) =>
    props.variant === 'ghost' &&
    css`
      background-color: var(--black-shade-5);
      color: var(--black-color-default);
      border: none;

      &:hover:not(:disabled) {
        background-color: var(--black-shade-3);
      }

      &:focus {
        box-shadow: 0 0 0 2px var(--black-shade-3);
      }
    `}

  /* Size styles */
  ${(props) =>
    props.size === '1' &&
    css`
      padding: 6px 12px;
    `}

  ${(props) =>
    props.size === '2' &&
    css`
      padding: 12px 16px;
    `}

  ${(props) =>
    props.size === '3' &&
    css`
      padding: 12px 24px;
    `}
`
