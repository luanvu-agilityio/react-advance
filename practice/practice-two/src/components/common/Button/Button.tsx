'use client'

import { type ButtonHTMLAttributes, type ReactNode } from 'react'
import { Button } from '@radix-ui/themes'
// Button variant types
type ButtonVariant = 'solid' | 'soft' | 'outline' | 'ghost'
type ButtonSize = '1' | '2' | '3'

// Button props interface
interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant
  size?: ButtonSize
  children: ReactNode
  onClick: () => void
}

// Button component
const BaseButton = ({
  variant = 'solid',
  size = '2',
  children,
  onClick,
}: ButtonProps) => {
  return (
    <Button variant={variant} size={size} onClick={onClick}>
      {children}
    </Button>
  )
}

export default BaseButton
export type { ButtonProps, ButtonVariant, ButtonSize }
