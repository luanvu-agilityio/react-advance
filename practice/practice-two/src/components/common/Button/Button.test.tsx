import { render, screen, fireEvent } from '@testing-library/react'
import Button from './Button'

describe('Button Snapshot Tests', () => {
  // Mock onClick handler for all snapshot tests
  const noop = () => {}

  test('renders all button variants and sizes correctly', () => {
    // Use React Testing Library for snapshots

    // Variants
    const { container: solidContainer } = render(
      <Button variant="solid" onClick={noop}>
        Solid Button
      </Button>
    )
    const { container: softContainer } = render(
      <Button variant="soft" onClick={noop}>
        Soft Button
      </Button>
    )
    const { container: outlineContainer } = render(
      <Button variant="outline" onClick={noop}>
        Outline Button
      </Button>
    )
    const { container: ghostContainer } = render(
      <Button variant="ghost" onClick={noop}>
        Ghost Button
      </Button>
    )

    // Sizes
    const { container: smallContainer } = render(
      <Button size="1" onClick={noop}>
        Small Button
      </Button>
    )
    const { container: mediumContainer } = render(
      <Button size="2" onClick={noop}>
        Medium Button
      </Button>
    )
    const { container: largeContainer } = render(
      <Button size="3" onClick={noop}>
        Large Button
      </Button>
    )

    // States
    const { container: disabledContainer } = render(
      <Button disabled onClick={noop}>
        Disabled Button
      </Button>
    )

    // With icon
    const { container: iconContainer } = render(
      <Button onClick={noop}>
        <span>üîç</span> Search
      </Button>
    )

    // Take snapshots
    expect(solidContainer.firstChild).toMatchSnapshot('solid-button')
    expect(softContainer.firstChild).toMatchSnapshot('soft-button')
    expect(outlineContainer.firstChild).toMatchSnapshot('outline-button')
    expect(ghostContainer.firstChild).toMatchSnapshot('ghost-button')
    expect(smallContainer.firstChild).toMatchSnapshot('small-button')
    expect(mediumContainer.firstChild).toMatchSnapshot('medium-button')
    expect(largeContainer.firstChild).toMatchSnapshot('large-button')
    expect(disabledContainer.firstChild).toMatchSnapshot('disabled-button')
    expect(iconContainer.firstChild).toMatchSnapshot('icon-button')
  })
})

describe('Button Interactive Tests', () => {
  test('calls onClick handler when clicked', () => {
    const handleClick = jest.fn()
    render(<Button onClick={handleClick}>Click Me</Button>)

    const button = screen.getByRole('button', { name: /click me/i })
    fireEvent.click(button)

    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  test('renders with different variants correctly', () => {
    const noop = jest.fn()

    // Test each variant in separate renders
    const { getByText: getSolidText } = render(
      <Button variant="solid" onClick={noop}>
        Solid
      </Button>
    )
    expect(getSolidText('Solid')).toBeInTheDocument()

    const { getByText: getSoftText } = render(
      <Button variant="soft" onClick={noop}>
        Soft
      </Button>
    )
    expect(getSoftText('Soft')).toBeInTheDocument()

    const { getByText: getOutlineText } = render(
      <Button variant="outline" onClick={noop}>
        Outline
      </Button>
    )
    expect(getOutlineText('Outline')).toBeInTheDocument()

    const { getByText: getGhostText } = render(
      <Button variant="ghost" onClick={noop}>
        Ghost
      </Button>
    )
    expect(getGhostText('Ghost')).toBeInTheDocument()
  })

  test('renders with different sizes correctly', () => {
    const noop = jest.fn()

    const { getByText: getSmallText } = render(
      <Button size="1" onClick={noop}>
        Small
      </Button>
    )
    expect(getSmallText('Small')).toBeInTheDocument()

    const { getByText: getMediumText } = render(
      <Button size="2" onClick={noop}>
        Medium
      </Button>
    )
    expect(getMediumText('Medium')).toBeInTheDocument()

    const { getByText: getLargeText } = render(
      <Button size="3" onClick={noop}>
        Large
      </Button>
    )
    expect(getLargeText('Large')).toBeInTheDocument()
  })
})
