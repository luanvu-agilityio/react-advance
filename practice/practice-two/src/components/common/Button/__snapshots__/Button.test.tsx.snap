// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot Tests renders all button variants and sizes correctly: disabled-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
  data-accent-color=""
>
  Disabled Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: ghost-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-ghost rt-Button"
  data-accent-color=""
>
  Ghost Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: icon-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
  data-accent-color=""
>
  <span>
    üîç
  </span>
   Search
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: large-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-3 rt-variant-solid rt-Button"
  data-accent-color=""
>
  Large Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: medium-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
  data-accent-color=""
>
  Medium Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: outline-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-outline rt-Button"
  data-accent-color=""
>
  Outline Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: small-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-1 rt-variant-solid rt-Button"
  data-accent-color=""
>
  Small Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: soft-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
  data-accent-color=""
>
  Soft Button
</button>
`;

exports[`Button Snapshot Tests renders all button variants and sizes correctly: solid-button 1`] = `
<button
  class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
  data-accent-color=""
>
  Solid Button
</button>
`;
